Ultimate Advanced Uber Super Mega Filmorate 2.0

Dvoychenkov opened the pull request.

EvgenyAlexandrovichM committed:
Добавлена реализация удаления пользователей и фильмов в соответствующие сервисы, реализованы эндпоинты к новому функционалу.
Добавлены тесты нового функционала

Commit:
Добавление режиссёров в фильмы.

Commit:
Добавление режиссёров в фильмы. code style

Dvoychenkov committed:
Merge pull request #5 from Dvoychenkov/add-remove-endpoint

Добавлена реализация удаления пользователей и фильмов в соответствующие сервисы, реализованы эндпоинты к новому функционалу.
Добавлены тесты нового функционала

Dvoychenkov committed:
- Дополнен init sql скрипт;

Dvoychenkov committed:
- Добавлена модель и DTO классы;

Dvoychenkov committed:
- Добавлен маппер для преобразования DTO-шек и модели;

Dvoychenkov committed:
- Добавлен интерфейс стораджа, его реализация для БД и Row маппер;

Dvoychenkov committed:
- Добавлен интерфейс сервиса и его базовая реализация;

Dvoychenkov committed:
- Добавлен контроллер;

Commit:
Добавление режиссёров в фильмы. code style

Commit:
Merge branch 'develop' into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/services/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/test/java/ru/yandex/practicum/filmorate/storage/FilmDbStorageTest.java

Commit:
Добавление режиссёров в фильмы. Решение конфликтов

Commit:
Добавление режиссёров в фильмы. Удаление режисеров

Commit:
Добавлена пустая строка в FilmService.

Commit:
Удален лишний импорт.

Commit:
Удалены лишние импорты.

Commit:
изменения, чтобы проходили тесты.

Commit:
Сделаны исправления после код-ревью.

rabarcre committed:
Добавлена реализация системы рекомендаций

Dvoychenkov committed:
Merge pull request #9 from Dvoychenkov/add-reviews

Добавление отзывов

Dvoychenkov committed:
Merge branch 'develop' into add-director

Commit:
Добавлен UnsupportedOperationException.

Dvoychenkov committed:
- Дополнен init sql скрипт;

Dvoychenkov committed:
- Добавлена модель и сопутствующие enum;

Dvoychenkov committed:
- Добавлен интерфейс стораджа, его реализация для БД и Row маппер;

Dvoychenkov committed:
- Добавлен интерфейс сервиса и его базовая реализация;

Dvoychenkov committed:
- Дополнены сущесствующие сервисы вызовом метода добавления события в фид;

Dvoychenkov committed:
- Дополнен контролер новым эндпоинтом;

Dvoychenkov committed:
- Поправлен и немного отрефакторен тестик валидации модели юзера в связи с правкой контроллера;

rabarcre committed:
Изменения связанные с ревью

Dvoychenkov committed:
Merge pull request #8 from Dvoychenkov/add-director

Добавление режиссёров в фильмы.

Dvoychenkov committed:
Merge branch 'develop' into add-feed

Dvoychenkov committed:
Merge pull request #11 from Dvoychenkov/add-feed

Функциональность «Лента событий»

Dvoychenkov committed:
- Удалено всё по теме отдельной сущности Recomendation - для такой задачи в этом нет необходимости;
- Дополнен интерфейс FilmStorage и добавлена реализация метода в FilmDbStorage + заглушка в InMemoryFilmStorage;
- Дополнен сервис и контроллер User (помещаем сюда, а не в Film классы согласно SRP);
- Удалено тело юнит-теста shouldNormalizeEmptyName, поскольку в develop ветке более актуальный вариант;

Dvoychenkov committed:
Merge branch 'develop' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/services/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/test/java/ru/yandex/practicum/filmorate/validation/UserValidationTest.java

Dvoychenkov committed:
- Фикс и рефакторинг юнит-теста;

Dvoychenkov committed:
- checkstyle fix;

Dvoychenkov committed:
Merge pull request #10 from Dvoychenkov/add-recommendations

Добавлена реализация системы рекомендаций

rabarcre committed:
Добавлена функциональность «Общие фильмы»

Commit:
получение популярных фильмов по жанру и году

Dvoychenkov committed:
- Добавлены минорные правки;

Dvoychenkov committed:
Merge pull request #12 from Dvoychenkov/add-common-films

Добавлена функциональность «Общие фильмы»

Commit:
Merge branch 'develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/services/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

Dvoychenkov committed:
Merge branch 'develop' into add-most-populars

Dvoychenkov committed:
- checkstyle fix;

Commit:
решение конфликтов

Commit:
Merge remote-tracking branch 'origin/add-most-populars' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/services/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

Commit:
решение конфликтов

Commit:
решение конфликтов

Dvoychenkov committed:
Merge pull request #16 from Dvoychenkov/add-most-populars

получение популярных фильмов по жанру и году

EvgenyAlexandrovichM committed:
Добавлен функционал поиска фильмов по подстроке, добавлен интерфейс FilmService.java

EvgenyAlexandrovichM committed:
Добавлен функционал поиска фильмов по подстроке, добавлен интерфейс FilmService.java

EvgenyAlexandrovichM committed:
Добавлен функционал поиска фильмов по подстроке, добавлен интерфейс FilmService.java

EvgenyAlexandrovichM committed:
Merge remote-tracking branch 'origin/add-search' into add-search

EvgenyAlexandrovichM committed:
Revert "Добавлен функционал поиска фильмов по подстроке, добавлен интерфейс FilmService.java"

This reverts commit 38c2c652f97e4261db29ee70a52409f870f7595d.

EvgenyAlexandrovichM committed:
Кодстайл

Dvoychenkov committed:
- Добавлен конвертер для валидации query параметра сортировки;
- Добавлен enum с типами сортировки;
- Поправлена реализация метода searchFilms в соответствии с новым подходом валидации входных данных;
- Поправлена сигнатура метода searchFilms контроллера в соответствие с новым подходом;
- Поправлено форматирование по мелочи;

Dvoychenkov committed:
Merge pull request #18 from Dvoychenkov/add-search

Add search

Dvoychenkov committed:
- Фикс мелкого бага согласно результатам тестов коллекции постмана;

Dvoychenkov committed:
- Фикс бага c добалвением событий в аудит согласно результатам тестов коллекции постмана + нано-рефакторинг;

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -51,14 +54,43 @@ public void addLike(@PathVariable Long filmId, @PathVariable Long userId) {
+
+    @GetMapping("/director/{directorId}")
+    public Collection<FilmDto> getDirectorFilms(@PathVariable Long directorId, @RequestParam SortOption sortBy) {
+        if (sortBy == SortOption.YEAR) {
```
Логику if/if else лучше вынести в сервис

Dvoychenkov replied:
Вынес

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/converters/StringToSortOptionConverter.java:
```diff
@@ -0,0 +1,13 @@
+public class StringToSortOptionConverter implements Converter<String, SortOption> {
+    @Override
+    public SortOption convert(String source) {
+        return SortOption.valueOf(source.toUpperCase());
```
Тут бы тоже try/catch добавить как в `StringToSearchByFieldConverter`

Dvoychenkov replied:
Добавил

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/mapper/DirectorMapper.java:
```diff
@@ -0,0 +1,27 @@
+import ru.yandex.practicum.filmorate.model.Director;
+
+@Component
+@RequiredArgsConstructor
```
В классе нет полей, аннотация кажется лишней

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/DirectorServiceBase.java:
```diff
@@ -0,0 +1,68 @@
+    public DirectorDto create(NewDirectorRequest newRequestDirector) {
+        Director directorToCreate = directorMapper.mapToDirector(newRequestDirector);
+        Director createdDirector = directorStorage.add(directorToCreate);
+        if (createdDirector == null) throw new IllegalStateException("Не удалось сохранить данные для нового режиссера");
```
Лучше выбрасывать ошибку внутри `directorStorage.add()`

Dvoychenkov replied:
Готово, сделал в аналогичном ключе для всех стораджей

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/FilmServiceBase.java:
```diff
@@ -0,0 +1,181 @@
+    public FilmDto create(NewFilmRequest newRequestFilm) {
+        Film filmToCreate = filmMapper.mapToFilm(newRequestFilm);
+        Film createdFilm = filmStorage.add(filmToCreate);
+        if (createdFilm == null) throw new IllegalStateException("Не удалось сохранить данные для нового фильма");
```
throw лучше перенести в `filmStorage.add()`

Dvoychenkov replied:
готово

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/FilmServiceBase.java:
```diff
@@ -0,0 +1,181 @@
+            log.info("Пользователь {} уже ставил лайк фильму {}", userId, filmId);
+        }
+
+        feedService.addEvent(new FeedEvent(userId, FeedEventType.LIKE, FeedOperation.ADD, filmId));
```
Кажется если пользователь уже поставил лайк, то ивент не нужен (перенсти метод ` feedService.addEvent` в if)

Dvoychenkov replied:
В общем, так изначально и было, но если вернуть добавление в аудит обртно в if(), то postman тесты начинают валиться.
![2025-05-20 21 41 37 057](https://github.com/user-attachments/assets/0da601fc-6922-48c7-9603-2df750b92007)

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/FilmServiceBase.java:
```diff
@@ -0,0 +1,181 @@
+            log.info("Пользователь {} не ставил лайк фильму {}", userId, filmId);
+        }
+
+        feedService.addEvent(new FeedEvent(userId, FeedEventType.LIKE, FeedOperation.REMOVE, filmId));
```
Аналогичный комментарий

Dvoychenkov replied:
Здесь тесты валиться не начали, вернул обратно фиксирование в аудит как надо

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/FilmServiceBase.java:
```diff
@@ -0,0 +1,181 @@
+
+    @Override
+    public Collection<FilmDto> getTopFilmsByLikes(int filmsLimit, Integer genreId, Integer year) {
+        if (filmsLimit <= 0) {
```
Кажется можно провалидировать `filmsLimit`  на уровне контроллера

Dvoychenkov replied:
поправил

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/FilmServiceBase.java:
```diff
@@ -0,0 +1,181 @@
+
+    @Override
+    public Film getFilmOrThrow(Long id) {
+        if (id == null) throw new ValidationException("Некорректный ID фильма");
```
null можно провалидировать на уровне контроллера

Dvoychenkov replied:
добавил

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/services/ReviewServiceBase.java:
```diff
@@ -0,0 +1,145 @@
+    @Override
+    public Collection<ReviewDto> getAll(Long filmId, Integer reviewsCount) {
+        // Если кол-во не указано, то по дефолту берём 10
+        if (reviewsCount == null || reviewsCount <= 0) {
```
Параметры можно провалидировать на уровне контроллера

Dvoychenkov replied:
Поправил валидацию для всех параметров, т.е. из всех сервисов вынес в соответствующие контроллеры

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/storage/feed/FeedRowMapper.java:
```diff
@@ -0,0 +1,35 @@
+        } catch (IllegalArgumentException e) {
+            throw new SQLException("Неизвестный тип события в БД: " + rs.getString("event_type"), e);
+        }
+        try {
```
Лучше объединить в один try catch

Dvoychenkov replied:
Объединил

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -121,8 +232,80 @@ public boolean removeLike(Long filmId, Long userId) {
+        }
+
+        String where = "WHERE " + String.join(" OR ", conditions);
+        String groupAndOrder = """
```
Можно вынести в поле класса

Dvoychenkov replied:
Скорректировал, заодно и сам метод поправил чуток

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -64,12 +65,53 @@ public boolean removeLike(Long filmId, Long userId) {
+        throw new UnsupportedOperationException();
+    }
+
+    // TODO улучшим?
```
`InMemory` реализацию можно не трогать)

Dvoychenkov replied:
Убрал TODO-шку)

Dvoychenkov committed:
- Убран неактуальный TODO;

Dvoychenkov committed:
Логика if/else перенесена:
- Проверка типа сортировки фильмов режиссёра вынесена из контроллера в сервис;
- Методы getDirectorFilmsSortedByLikes и getDirectorFilmsSortedByYears объединены в один getDirectorFilms;
- Отформатирован sql-запрос получения инфы о фильмах режиссёра по годам;

Dvoychenkov committed:
- Дополнена и отрефакторена проверка конвертеров входных параметров типа enum;

Dvoychenkov committed:
- Убрана лишняя аннотация;

Dvoychenkov committed:
- Выброс исключений для create-операций перенесён непосредственно в стораджи;

Dvoychenkov committed:
- Вернул запись в аудит об удалении лайка обратно. Для добавления оставил как есть - иначе падают тесты postman;

Dvoychenkov committed:
Рефакторинг и правка подхода к валидации:
- Добавлен кастомный валидатор для id-шников;
- Валидация параметров перенесена из сервисов в контроллеры;

Dvoychenkov committed:
- Отрефакторен и улучшен метод по построению запроса для поиска фильмов;
- Упрощена проверка ошибок в FeedRowMapper;

Dvoychenkov committed:
- Добавлено описание по выполненным таскам группового проекта;

Dvoychenkov commented:
@slavrikk Дополнил в README что было сделано по групповому проекту

Кстати по теме комментария:
>Есть несколько вложенных запросов, но в целом не критично. 

Изначально я сделал такой запрос по рекомендациям, но по какой-то причине при наличии CTE в запросе игнорировались его параметры, сходу победить не смог, поэтому было принято решение сделать альетрнативный рабочий вариант в силу ограниченности ресурсов в команде.

``` sql
/* Ищем топ юзеров с максимальным пересечением по лайкам */
WITH max_intersect_likes_users AS (
  select count(1) AS films_likes_cnt, other_likes.USER_ID AS other_user_id
  from FILMS_USERS_LIKES user_likes
  JOIN FILMS_USERS_LIKES other_likes ON other_likes.FILM_ID = user_likes.FILM_ID
  WHERE user_likes.USER_ID = 25
    AND other_likes.USER_ID != 25
  GROUP BY other_likes.USER_ID
)
/* Оставляем самого топового юзера, пусть будет "Рекомендатель" */
,top_likes_intersect_user AS (
  select other_user_id
  from max_intersect_likes_users
  ORDER BY films_likes_cnt DESC
  LIMIT 1
)
/* Получаем все фильмы Рекомендателя */
,other_user_liked_films AS (
  SELECT film_id
  FROM FILMS_USERS_LIKES
  WHERE user_id IN (SELECT other_user_id FROM top_likes_intersect_user)
)
/* Находим фильмы, которые лайкнул Рекомендатель, но не лайкнул целевой пользователь */
SELECT * FROM films
WHERE id IN (SELECT film_id FROM other_user_liked_films)
AND id NOT IN (select film_id from FILMS_USERS_LIKES WHERE USER_ID = 25);

```

Dvoychenkov commented:
Продублирую дополненную часть README сюда

##  Реализованные задачи в рамках группового спринта

### 1. Удаление фильмов и пользователей
Добавлена поддержка удаления по идентификатору:
-   `DELETE /users/{userId}` — удаление пользователя
-   `DELETE /films/{filmId}` — удаление фильма

---

### 2. Отзывы к фильмам (`/reviews`)
Реализована система отзывов с лайками и рейтингом полезности:
-   `POST /reviews` — создание
-   `PUT /reviews` — редактирование
-   `DELETE /reviews/{id}` — удаление
-   `GET /reviews/{id}` — получение по ID
-   `GET /reviews?filmId=&count=` — список по фильму
-   `PUT/DELETE /reviews/{id}/like/{userId}` — лайки
-   `PUT/DELETE /reviews/{id}/dislike/{userId}` — дизлайки

У отзыва есть:
-   текст (`content`)
-   тип (`isPositive`)
-   рейтинг (`useful`, увеличивается/уменьшается при голосах)

---

### 3. Режиссёры и сортировка фильмов (`/films/director`)
Добавлена поддержка работы с режиссёрами:
-   `GET /directors` — получение всех
-   `GET /directors/{id}` — получение по ID
-   `POST /directors` — создание
-   `PUT /directors` — редактирование
-   `DELETE /directors/{id}` — удаление

Фильмы можно сортировать:
-   по лайкам: `GET /films/director/{id}?sortBy=likes`
-   по годам: `GET /films/director/{id}?sortBy=year`

---

### 4. Лента событий (`/users/{id}/feed`)
Добавлена лента активности пользователя:
-   поддержка событий `LIKE`, `REVIEW`, `FRIEND`
-   операции `ADD`, `REMOVE`, `UPDATE` (обновление только для отзывов)
-   `GET /users/{id}/feed` возвращает хронологический список действий

---

### 5. Рекомендации фильмов (`/users/{id}/recommendations`)
Реализован простой алгоритм рекомендаций:
1. находит пользователя с похожими вкусами
2. рекомендует фильмы, которые он лайкал, а целевой пользователь — нет
-   `GET /users/{id}/recommendations`

---

### 6. Общие фильмы (`/films/common`)
Добавлен эндпоинт:
-   `GET /films/common?userId=&friendId=`
-   возвращает фильмы, лайкнутые обоими пользователями
-   отсортировано по количеству лайков (популярности)

---

### 7. Популярные фильмы по фильтрам (`/films/popular`)
-   `GET /films/popular?count=&genreId=&year=`
-   поддержка фильтрации по жанру и году
-   отсортировано по количеству лайков

---

### 8. Поиск фильмов (`/films/search`)
Реализован поиск по подстроке:
-   `GET /films/search?query=...&by=title,director`
-   ищет по названию фильма и/или имени режиссёра
-   сортировка по популярности

---

### Произведённый рефакторинг и улучшения
-   Все ID валидируются через кастомную аннотацию `@IdValid("userId" / "filmId" / ...)`
-   Используются `@NotNull`, `@NotBlank`, `@Size`, `@Min` для параметров
-   Поддержка `enum`-значений через Spring `Converter` (`SortOption`, `SearchByField`)
-   В целом вся валидация вынесена в контроллеры
-   Выброс исключений для create операций перенесён в storage для каждого функционала
